compiler/AST/foralls.cpp

```diff
@@ -230,3 +234,763 @@ bool astUnderFI(const Expr* ast, ForallIntents* fi) {
+                                          new_Expr("'move'(%S, chpl__dynamicFastFollowCheck(%S))",    T2, iterRec),
+                                          new_Expr("'move'(%S, %S)", T2, gFalse)));
+    } else {
+      leadForLoop->insertAtTail("'move'(%S, chpl__staticFastFollowCheckZip(%S))", T1, iterRec);
+      leadForLoop->insertAtTail(new CondStmt(new SymExpr(T1),
+                                          new_Expr("'move'(%S, chpl__dynamicFastFollowCheckZip(%S))", T2, iterRec),
+                                          new_Expr("'move'(%S, %S)", T2, gFalse)));
+    }
+
+    SymbolMap map;
+    map.put(followIdx, fastFollowIdx);
+    BlockStmt* loopBodyForFast = loopBody->copy(&map);
+
+    fastFollowBlock = buildFollowLoop(iterRec,
+                                      leadIdxCopy,
+                                      fastFollowIter,
+                                      fastFollowIdx,
+                                      loopBodyForFast,
+                                      pfs,
+                                      true,
+                                      zippered);
+
+    leadForLoop->insertAtTail(new CondStmt(new SymExpr(T2), fastFollowBlock, followBlock));
+  } else {
+    leadForLoop->insertAtTail(followBlock);
+  }
+
+  // Must happen before any resolving ex. resolveBlockStmt below.
+  // Otherwise functions defined within the forall body are not visible.
+  // Ex. functions/vass/ref-intent-bug-2big.chpl
+  pfs->masterBlock()->insertAtTail(leadForLoop);
+
+  pfs->insertBefore(resultBlock);
+  normalize(toNormalize); // requires inTree()
+  resolveBlockStmt(resultBlock);
+  resultBlock->flattenAndRemove();
+}
+
+Expr* resolveParallelIteratorAndForallIntents(ForallStmt* pfs, SymExpr* origSE)
+{
+  // We only get here for origSE==firstIterExpr() .
+  // If at that time there are other elements in iterExprs(), we remove them.
+  INT_ASSERT(origSE == pfs->firstIterExpr());
+
+  CallExpr* iterCall = buildForallParIterCall(pfs, origSE);
+
+  // So we know where iterCall is.
+  INT_ASSERT(iterCall == pfs->firstIterExpr());
+  INT_ASSERT(!origSE->inTree());
+
+  bool gotSA = findStandaloneOrLeader(pfs, iterCall);
+
+  if (gotSA) {
+    // Former buildForallLoopStmt() added it to leadIdxCopy, not saIdxCopy.
+    // If we retain this flag, 'filename' gets double-delete'd and crashes:
+    //   test/modules/standard/FileSystem/filerator/bradc/findfiles-par.chpl
+    // Todo: why?
+    pfs->parIdxCopyVar()->removeFlag(FLAG_INSERT_AUTO_DESTROY);
+    // NB former buildForallLoopStmt() set this flag on saIdx, not on leadIdx.
+    // We better set it here - it's needed in setConstFlagsAndCheckUponMove().
+    pfs->parIdxVar()->addFlag(FLAG_INDEX_OF_INTEREST);
+    
+  } else    // vass: just curious; remove this "else" clause after testing
+    // NB former buildForallLoopStmt() set this flag on saIdx, not on leadIdx.
+    INT_ASSERT(!pfs->parIdxVar()->hasFlag(FLAG_INDEX_OF_INTEREST));
+
```

Comment:
I will remove the above assertion in the final version - it does not feel useful.